File: OTCDerivativesCalculatorModule/Project_Cpp/lib_static/FpmlSerialized/GenClass/fpml-eq-shared-5-4/DirectionalLegUnderlyer.cpp
// DirectionalLegUnderlyer.cpp 
#include "DirectionalLegUnderlyer.hpp"
#ifdef ConsolePrint
    #include <iostream>
#endif 

namespace FpmlSerialized {

DirectionalLegUnderlyer::DirectionalLegUnderlyer(TiXmlNode* xmlNode)
: DirectionalLeg(xmlNode)
{
    #ifdef ConsolePrint
        std::string initialtap_ = FileManager::instance().tap_;
        FileManager::instance().tap_.append("   ");
    #endif 
   //underlyerNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* underlyerNode = xmlNode->FirstChildElement("underlyer");

   if(underlyerNode){underlyerIsNull_ = false;}
   else{underlyerIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- underlyerNode , address : " << underlyerNode << std::endl;
   #endif
   if(underlyerNode)
   {
      if(underlyerNode->Attribute("href") || underlyerNode->Attribute("id"))
      {
          if(underlyerNode->Attribute("id"))
          {
             underlyerIDRef_ = underlyerNode->Attribute("id");
             underlyer_ = boost::shared_ptr<Underlyer>(new Underlyer(underlyerNode));
             underlyer_->setID(underlyerIDRef_);
             IDManager::instance().setID(underlyerIDRef_,underlyer_);
          }
          else if(underlyerNode->Attribute("href")) { underlyerIDRef_ = underlyerNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { underlyer_ = boost::shared_ptr<Underlyer>(new Underlyer(underlyerNode));}
   }

   //settlementTypeNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* settlementTypeNode = xmlNode->FirstChildElement("settlementType");

   if(settlementTypeNode){settlementTypeIsNull_ = false;}
   else{settlementTypeIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- settlementTypeNode , address : " << settlementTypeNode << std::endl;
   #endif
   if(settlementTypeNode)
   {
      if(settlementTypeNode->Attribute("href") || settlementTypeNode->Attribute("id"))
      {
          if(settlementTypeNode->Attribute("id"))
          {
             settlementTypeIDRef_ = settlementTypeNode->Attribute("id");
             settlementType_ = boost::shared_ptr<SettlementTypeEnum>(new SettlementTypeEnum(settlementTypeNode));
             settlementType_->setID(settlementTypeIDRef_);
             IDManager::instance().setID(settlementTypeIDRef_,settlementType_);
          }
          else if(settlementTypeNode->Attribute("href")) { settlementTypeIDRef_ = settlementTypeNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { settlementType_ = boost::shared_ptr<SettlementTypeEnum>(new SettlementTypeEnum(settlementTypeNode));}
   }

   //settlementDateNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* settlementDateNode = xmlNode->FirstChildElement("settlementDate");

   if(settlementDateNode){settlementDateIsNull_ = false;}
   else{settlementDateIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- settlementDateNode , address : " << settlementDateNode << std::endl;
   #endif
   if(settlementDateNode)
   {
      if(settlementDateNode->Attribute("href") || settlementDateNode->Attribute("id"))
      {
          if(settlementDateNode->Attribute("id"))
          {
             settlementDateIDRef_ = settlementDateNode->Attribute("id");
             settlementDate_ = boost::shared_ptr<AdjustableOrRelativeDate>(new AdjustableOrRelativeDate(settlementDateNode));
             settlementDate_->setID(settlementDateIDRef_);
             IDManager::instance().setID(settlementDateIDRef_,settlementDate_);
          }
          else if(settlementDateNode->Attribute("href")) { settlementDateIDRef_ = settlementDateNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { settlementDate_ = boost::shared_ptr<AdjustableOrRelativeDate>(new AdjustableOrRelativeDate(settlementDateNode));}
   }

   //settlementAmountNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* settlementAmountNode = xmlNode->FirstChildElement("settlementAmount");

   if(settlementAmountNode){settlementAmountIsNull_ = false;}
   else{settlementAmountIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- settlementAmountNode , address : " << settlementAmountNode << std::endl;
   #endif
   if(settlementAmountNode)
   {
      if(settlementAmountNode->Attribute("href") || settlementAmountNode->Attribute("id"))
      {
          if(settlementAmountNode->Attribute("id"))
          {
             settlementAmountIDRef_ = settlementAmountNode->Attribute("id");
             settlementAmount_ = boost::shared_ptr<Money>(new Money(settlementAmountNode));
             settlementAmount_->setID(settlementAmountIDRef_);
             IDManager::instance().setID(settlementAmountIDRef_,settlementAmount_);
          }
          else if(settlementAmountNode->Attribute("href")) { settlementAmountIDRef_ = settlementAmountNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { settlementAmount_ = boost::shared_ptr<Money>(new Money(settlementAmountNode));}
   }

   //settlementCurrencyNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* settlementCurrencyNode = xmlNode->FirstChildElement("settlementCurrency");

   if(settlementCurrencyNode){settlementCurrencyIsNull_ = false;}
   else{settlementCurrencyIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- settlementCurrencyNode , address : " << settlementCurrencyNode << std::endl;
   #endif
   if(settlementCurrencyNode)
   {
      if(settlementCurrencyNode->Attribute("href") || settlementCurrencyNode->Attribute("id"))
      {
          if(settlementCurrencyNode->Attribute("id"))
          {
             settlementCurrencyIDRef_ = settlementCurrencyNode->Attribute("id");
             settlementCurrency_ = boost::shared_ptr<Currency>(new Currency(settlementCurrencyNode));
             settlementCurrency_->setID(settlementCurrencyIDRef_);
             IDManager::instance().setID(settlementCurrencyIDRef_,settlementCurrency_);
          }
          else if(settlementCurrencyNode->Attribute("href")) { settlementCurrencyIDRef_ = settlementCurrencyNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { settlementCurrency_ = boost::shared_ptr<Currency>(new Currency(settlementCurrencyNode));}
   }

   //fxFeatureNode ----------------------------------------------------------------------------------------------------------------------
   TiXmlElement* fxFeatureNode = xmlNode->FirstChildElement("fxFeature");

   if(fxFeatureNode){fxFeatureIsNull_ = false;}
   else{fxFeatureIsNull_ = true;}

   #ifdef ConsolePrint
      FileManager::instance().outFile_ << FileManager::instance().tap_.c_str() << "- fxFeatureNode , address : " << fxFeatureNode << std::endl;
   #endif
   if(fxFeatureNode)
   {
      if(fxFeatureNode->Attribute("href") || fxFeatureNode->Attribute("id"))
      {
          if(fxFeatureNode->Attribute("id"))
          {
             fxFeatureIDRef_ = fxFeatureNode->Attribute("id");
             fxFeature_ = boost::shared_ptr<FxFeature>(new FxFeature(fxFeatureNode));
             fxFeature_->setID(fxFeatureIDRef_);
             IDManager::instance().setID(fxFeatureIDRef_,fxFeature_);
          }
          else if(fxFeatureNode->Attribute("href")) { fxFeatureIDRef_ = fxFeatureNode->Attribute("href");}
          else { QL_FAIL("id or href error"); }
      }
      else { fxFeature_ = boost::shared_ptr<FxFeature>(new FxFeature(fxFeatureNode));}
   }

    #ifdef ConsolePrint
        FileManager::instance().tap_ = initialtap_;
    #endif 
}
boost::shared_ptr<Underlyer> DirectionalLegUnderlyer::getUnderlyer()
{
   if(!this->underlyerIsNull_){
        if(underlyerIDRef_ != ""){
             return boost::shared_static_cast<Underlyer>(IDManager::instance().getID(underlyerIDRef_));
        }else{
             return this->underlyer_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<Underlyer>();
   }
}
boost::shared_ptr<SettlementTypeEnum> DirectionalLegUnderlyer::getSettlementType()
{
   if(!this->settlementTypeIsNull_){
        if(settlementTypeIDRef_ != ""){
             return boost::shared_static_cast<SettlementTypeEnum>(IDManager::instance().getID(settlementTypeIDRef_));
        }else{
             return this->settlementType_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<SettlementTypeEnum>();
   }
}
boost::shared_ptr<AdjustableOrRelativeDate> DirectionalLegUnderlyer::getSettlementDate()
{
   if(!this->settlementDateIsNull_){
        if(settlementDateIDRef_ != ""){
             return boost::shared_static_cast<AdjustableOrRelativeDate>(IDManager::instance().getID(settlementDateIDRef_));
        }else{
             return this->settlementDate_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<AdjustableOrRelativeDate>();
   }
}
boost::shared_ptr<Money> DirectionalLegUnderlyer::getSettlementAmount()
{
   if(!this->settlementAmountIsNull_){
        if(settlementAmountIDRef_ != ""){
             return boost::shared_static_cast<Money>(IDManager::instance().getID(settlementAmountIDRef_));
        }else{
             return this->settlementAmount_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<Money>();
   }
}
boost::shared_ptr<Currency> DirectionalLegUnderlyer::getSettlementCurrency()
{
   if(!this->settlementCurrencyIsNull_){
        if(settlementCurrencyIDRef_ != ""){
             return boost::shared_static_cast<Currency>(IDManager::instance().getID(settlementCurrencyIDRef_));
        }else{
             return this->settlementCurrency_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<Currency>();
   }
}
boost::shared_ptr<FxFeature> DirectionalLegUnderlyer::getFxFeature()
{
   if(!this->fxFeatureIsNull_){
        if(fxFeatureIDRef_ != ""){
             return boost::shared_static_cast<FxFeature>(IDManager::instance().getID(fxFeatureIDRef_));
        }else{
             return this->fxFeature_;
        }
   }else
   {
      QL_FAIL("null Ptr");
      return boost::shared_ptr<FxFeature>();
   }
}
}

File: OTCDerivativesCalculatorModule/Project_CSharp/XmlBuilder/XmlSerializedClass/fpml-eq-shared-5-4/DirectionalLegUnderlyer.cs
using System;
using System.Text;
using System.Collections.Generic;
using System.Xml;

namespace FpmlSerializedCSharp
{
    public class DirectionalLegUnderlyer : DirectionalLeg
    {
        public DirectionalLegUnderlyer(XmlNode xmlNode)
        : base(xmlNode)
        {
            XmlNode underlyerNode = xmlNode.SelectSingleNode("underlyer");
            
            if (underlyerNode != null)
            {
                if (underlyerNode.Attributes["href"] != null || underlyerNode.Attributes["id"] != null) 
                {
                    if (underlyerNode.Attributes["id"] != null) 
                    {
                        underlyerIDRef_ = underlyerNode.Attributes["id"].Value;
                        Underlyer ob = new Underlyer(underlyerNode);
                        IDManager.SetID(underlyerIDRef_, ob);
                    }
                    else if (underlyerNode.Attributes["href"] != null)
                    {
                        underlyerIDRef_ = underlyerNode.Attributes["href"].Value;
                    }
                    else
                    {
                        underlyer_ = new Underlyer(underlyerNode);
                    }
                }
                else
                {
                    underlyer_ = new Underlyer(underlyerNode);
                }
            }
            
        
            XmlNode settlementTypeNode = xmlNode.SelectSingleNode("settlementType");
            
            if (settlementTypeNode != null)
            {
                if (settlementTypeNode.Attributes["href"] != null || settlementTypeNode.Attributes["id"] != null) 
                {
                    if (settlementTypeNode.Attributes["id"] != null) 
                    {
                        settlementTypeIDRef_ = settlementTypeNode.Attributes["id"].Value;
                        SettlementTypeEnum ob = new SettlementTypeEnum(settlementTypeNode);
                        IDManager.SetID(settlementTypeIDRef_, ob);
                    }
                    else if (settlementTypeNode.Attributes["href"] != null)
                    {
                        settlementTypeIDRef_ = settlementTypeNode.Attributes["href"].Value;
                    }
                    else
                    {
                        settlementType_ = new SettlementTypeEnum(settlementTypeNode);
                    }
                }
                else
                {
                    settlementType_ = new SettlementTypeEnum(settlementTypeNode);
                }
            }
            
        
            XmlNode settlementDateNode = xmlNode.SelectSingleNode("settlementDate");
            
            if (settlementDateNode != null)
            {
                if (settlementDateNode.Attributes["href"] != null || settlementDateNode.Attributes["id"] != null) 
                {
                    if (settlementDateNode.Attributes["id"] != null) 
                    {
                        settlementDateIDRef_ = settlementDateNode.Attributes["id"].Value;
                        AdjustableOrRelativeDate ob = new AdjustableOrRelativeDate(settlementDateNode);
                        IDManager.SetID(settlementDateIDRef_, ob);
                    }
                    else if (settlementDateNode.Attributes["href"] != null)
                    {
                        settlementDateIDRef_ = settlementDateNode.Attributes["href"].Value;
                    }
                    else
                    {
                        settlementDate_ = new AdjustableOrRelativeDate(settlementDateNode);
                    }
                }
                else
                {
                    settlementDate_ = new AdjustableOrRelativeDate(settlementDateNode);
                }
            }
            
        
            XmlNode settlementAmountNode = xmlNode.SelectSingleNode("settlementAmount");
            
            if (settlementAmountNode != null)
            {
                if (settlementAmountNode.Attributes["href"] != null || settlementAmountNode.Attributes["id"] != null) 
                {
                    if (settlementAmountNode.Attributes["id"] != null) 
                    {
                        settlementAmountIDRef_ = settlementAmountNode.Attributes["id"].Value;
                        Money ob = new Money(settlementAmountNode);
                        IDManager.SetID(settlementAmountIDRef_, ob);
                    }
                    else if (settlementAmountNode.Attributes["href"] != null)
                    {
                        settlementAmountIDRef_ = settlementAmountNode.Attributes["href"].Value;
                    }
                    else
                    {
                        settlementAmount_ = new Money(settlementAmountNode);
                    }
                }
                else
                {
                    settlementAmount_ = new Money(settlementAmountNode);
                }
            }
            
        
            XmlNode settlementCurrencyNode = xmlNode.SelectSingleNode("settlementCurrency");
            
            if (settlementCurrencyNode != null)
            {
                if (settlementCurrencyNode.Attributes["href"] != null || settlementCurrencyNode.Attributes["id"] != null) 
                {
                    if (settlementCurrencyNode.Attributes["id"] != null) 
                    {
                        settlementCurrencyIDRef_ = settlementCurrencyNode.Attributes["id"].Value;
                        Currency ob = new Currency(settlementCurrencyNode);
                        IDManager.SetID(settlementCurrencyIDRef_, ob);
                    }
                    else if (settlementCurrencyNode.Attributes["href"] != null)
                    {
                        settlementCurrencyIDRef_ = settlementCurrencyNode.Attributes["href"].Value;
                    }
                    else
                    {
                        settlementCurrency_ = new Currency(settlementCurrencyNode);
                    }
                }
                else
                {
                    settlementCurrency_ = new Currency(settlementCurrencyNode);
                }
            }
            
        
            XmlNode fxFeatureNode = xmlNode.SelectSingleNode("fxFeature");
            
            if (fxFeatureNode != null)
            {
                if (fxFeatureNode.Attributes["href"] != null || fxFeatureNode.Attributes["id"] != null) 
                {
                    if (fxFeatureNode.Attributes["id"] != null) 
                    {
                        fxFeatureIDRef_ = fxFeatureNode.Attributes["id"].Value;
                        FxFeature ob = new FxFeature(fxFeatureNode);
                        IDManager.SetID(fxFeatureIDRef_, ob);
                    }
                    else if (fxFeatureNode.Attributes["href"] != null)
                    {
                        fxFeatureIDRef_ = fxFeatureNode.Attributes["href"].Value;
                    }
                    else
                    {
                        fxFeature_ = new FxFeature(fxFeatureNode);
                    }
                }
                else
                {
                    fxFeature_ = new FxFeature(fxFeatureNode);
                }
            }
            
        
        }
        
    
        #region Underlyer_
        private Underlyer underlyer_;
        public Underlyer Underlyer_
        {
            get
            {
                if (this.underlyer_ != null)
                {
                    return this.underlyer_; 
                }
                else if (this.underlyerIDRef_ != null)
                {
                    underlyer_ = IDManager.getID(underlyerIDRef_) as Underlyer;
                    return this.underlyer_; 
                }
                else
                {
                      return this.underlyer_; 
                }
            }
            set
            {
                if (this.underlyer_ != value)
                {
                    this.underlyer_ = value;
                }
            }
        }
        #endregion
        
        public string underlyerIDRef_ { get; set; }
        #region SettlementType_
        private SettlementTypeEnum settlementType_;
        public SettlementTypeEnum SettlementType_
        {
            get
            {
                if (this.settlementType_ != null)
                {
                    return this.settlementType_; 
                }
                else if (this.settlementTypeIDRef_ != null)
                {
                    settlementType_ = IDManager.getID(settlementTypeIDRef_) as SettlementTypeEnum;
                    return this.settlementType_; 
                }
                else
                {
                      return this.settlementType_; 
                }
            }
            set
            {
                if (this.settlementType_ != value)
                {
                    this.settlementType_ = value;
                }
            }
        }
        #endregion
        
        public string settlementTypeIDRef_ { get; set; }
        #region SettlementDate_
        private AdjustableOrRelativeDate settlementDate_;
        public AdjustableOrRelativeDate SettlementDate_
        {
            get
            {
                if (this.settlementDate_ != null)
                {
                    return this.settlementDate_; 
                }
                else if (this.settlementDateIDRef_ != null)
                {
                    settlementDate_ = IDManager.getID(settlementDateIDRef_) as AdjustableOrRelativeDate;
                    return this.settlementDate_; 
                }
                else
                {
                      return this.settlementDate_; 
                }
            }
            set
            {
                if (this.settlementDate_ != value)
                {
                    this.settlementDate_ = value;
                }
            }
        }
        #endregion
        
        public string settlementDateIDRef_ { get; set; }
        #region SettlementAmount_
        private Money settlementAmount_;
        public Money SettlementAmount_
        {
            get
            {
                if (this.settlementAmount_ != null)
                {
                    return this.settlementAmount_; 
                }
                else if (this.settlementAmountIDRef_ != null)
                {
                    settlementAmount_ = IDManager.getID(settlementAmountIDRef_) as Money;
                    return this.settlementAmount_; 
                }
                else
                {
                      return this.settlementAmount_; 
                }
            }
            set
            {
                if (this.settlementAmount_ != value)
                {
                    this.settlementAmount_ = value;
                }
            }
        }
        #endregion
        
        public string settlementAmountIDRef_ { get; set; }
        #region SettlementCurrency_
        private Currency settlementCurrency_;
        public Currency SettlementCurrency_
        {
            get
            {
                if (this.settlementCurrency_ != null)
                {
                    return this.settlementCurrency_; 
                }
                else if (this.settlementCurrencyIDRef_ != null)
                {
                    settlementCurrency_ = IDManager.getID(settlementCurrencyIDRef_) as Currency;
                    return this.settlementCurrency_; 
                }
                else
                {
                      return this.settlementCurrency_; 
                }
            }
            set
            {
                if (this.settlementCurrency_ != value)
                {
                    this.settlementCurrency_ = value;
                }
            }
        }
        #endregion
        
        public string settlementCurrencyIDRef_ { get; set; }
        #region FxFeature_
        private FxFeature fxFeature_;
        public FxFeature FxFeature_
        {
            get
            {
                if (this.fxFeature_ != null)
                {
                    return this.fxFeature_; 
                }
                else if (this.fxFeatureIDRef_ != null)
                {
                    fxFeature_ = IDManager.getID(fxFeatureIDRef_) as FxFeature;
                    return this.fxFeature_; 
                }
                else
                {
                      return this.fxFeature_; 
                }
            }
            set
            {
                if (this.fxFeature_ != value)
                {
                    this.fxFeature_ = value;
                }
            }
        }
        #endregion
        
        public string fxFeatureIDRef_ { get; set; }
        public string choiceStr_0;
        
    
        
    
    }
    
}

